### PR TITLE
Rich media support

### PR DESCRIPTION
**Pull request by hamchapman**

### CODE REVIEW

**Comment by hamchapman** [comment]
Still needs some cleanup.

Also, @hot-leaf-juice I'm getting a `NullPointerException` here:

```java
final int position = user.getCursors().get(roomId).getPosition();
```

Any ideas why? It basically says that I'm calling `getPosition()` on `null`

**Comment by hot-leaf-juice** [comment]
@hamchapman I suspect that there is no cursor set for `roomId`, so the `get` returns `null`. Where is that code? If that's in the library somewhere I should fix it to do a null check, otherwise I guess it's fair for us to leave that to the consumer?

File: chatkit/src/main/java/com/pusher/chatkit/CurrentUser.kt [hunk 1]
```diff
@@ -173,34 +183,114 @@ class CurrentUser(
...
+            text: String? = null,
+            attachment: GenericAttachment? = null,
             onCompleteListener: MessageSentListener,
             onErrorListener: ErrorListener
-    ){
-        val messageReq = MessageRequest(
+    ): Cancelable? {
+        var attachmentBody: AttachmentBody? = null
+
+        if (attachment != null) {
```
Comment (vivangkumar) [inline chatkit/src/main/java/com/pusher/chatkit/CurrentUser.kt]: We can use some async await magic here if you want to clean it up. Not that I am against this, but I reverted to this just to be able to debug that error we had.

**Comment by hamchapman** [comment]
Now I have at least the smallest grasp of how the coroutines stuff is supposed to work, it's actually pretty sweet.

Updated a couple of the functions in the code paths I had updated:

https://github.com/pusher/chatkit-android/pull/4/files#diff-e6bbe2d672f6bb996064847db66d2b7cR279
https://github.com/pusher/chatkit-android/pull/4/files#diff-e6bbe2d672f6bb996064847db66d2b7cR235

Also, this part of it still needs to be sorted (i.e. I have no idea what I'm doing with snapshots):

https://github.com/pusher/chatkit-android/pull/4/files#diff-1bdb8e35c53c9e3cf971ea14eb88b101R42

**Comment by hamchapman** [comment]
@vivangkumar and @hot-leaf-juice can you please take (another) look at it now that I've made the coroutine changes? 👍

File: app/src/main/java/com/pusher/chatkit/sample/ChatRoomActivity.java [hunk 2]
```diff
@@ -128,38 +161,194 @@ public void onError(Error error) {
...
+                File file = new File(filePath);
+                this.selectedFile = file;
+            } else {
+                Log.d("PC", "File path is null");
+            }
+        }
+    }
 
-    private void showNewMessageDialog() {
+    public final void throwUpThePicker() {
```
Comment (hot-leaf-juice) [inline app/src/main/java/com/pusher/chatkit/sample/ChatRoomActivity.java]: 🤮

File: chatkit/build.gradle [hunk 3]
```diff
@@ -33,6 +38,8 @@ dependencies {
         exclude group: 'com.android.support', module: 'support-annotations'
     })
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jre7:$kotlin_version"
-    api 'com.pusher:pusher-platform:0.0.1-SNAPSHOT' //Remote pusher platform
     api 'org.jetbrains.kotlinx:kotlinx-coroutines-core:0.21.2'
+//    api 'com.pusher:pusher-platform:0.0.1-SNAPSHOT' //Remote pusher platform
```
Comment (vivangkumar) [inline chatkit/build.gradle]: This bit really just needs to be uncommented and the line below removed. But we have to make sure we release the platform lib first and update the snapshot.

File: settings.gradle [hunk 4]
```diff
@@ -1,5 +1,3 @@
-include ':app', ':chatkit'
-
-
-
+include ':app', ':chatkit', ':pusher-platform-local'
```
Comment (vivangkumar) [inline settings.gradle]: also all of this needs to go

File: chatkit/src/main/java/com/pusher/chatkit/Result.kt [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package com.pusher.chatkit
+
+/**
+ * Code from https://github.com/michaelbull/kotlin-result
```
Comment (hot-leaf-juice) [inline chatkit/src/main/java/com/pusher/chatkit/Result.kt]: Can we include import this with gradle a al https://github.com/michaelbull/kotlin-result#installation?

Comment (hamchapman) [reply chatkit/src/main/java/com/pusher/chatkit/Result.kt]: Could do but that has a load more utility functions for interacting with the `Result` that we don't need.

Comment (hot-leaf-juice) [reply chatkit/src/main/java/com/pusher/chatkit/Result.kt]: fair enough. I'm a little bit fuzzy on whether a link to the lisence is enough to fulfil

> provided that the above copyright notice and this permission notice appear in all copies

but I'm sure I'm splitting hairs here 😛

**Pull request merged**